R/tag_query.R

```diff
@@ -0,0 +1,1601 @@
+#'
+#' ## Tag environments
+#'
+#' "Tag environments" are the building blocks to `tagQuery()`. When creating a
+#' `tagQuery()`, each tag object is converted to a tag environment. This
+#' conversion allows for element alterations to happen in place (pass by
+#' reference). Meaning that if a css class is added to each selected tag
+#' environment using `$addClass()` and the result of the method call is ignored,
+#' the selected tag environments within the tag query object will still contain
+#' the class addition.  The added class will exist when the tag query tag
+#' environment are converted back to standard tags objects with `$asTags()`.
+#'
+#' Tag environments also contain an extra field, `.$parent`. The `.$parent`
+#' value contains their parent tag environment. The top level tags supplied to
+#' `tagQuery()` will also have a shared parent element. (The shared parent
+#' element will have a `NULL` `.$parent` value.) This allows for performing
+#' sibling alterations at the top level of the tag query object.
+#'
+#' The set of tag environments in a pointing to each other within a tag query
+#' object can be thought of as a linked list while allowing for a "1 to many"
+#' parent to child relationship and up to 1 parent per child.
+#'
+#' ## Tag Query
+#'
+#' A `tagQuery()` behaves simliar to an R6 object (but a `tagQuery()` object is
+#' not implemented with `R6`). The `tagQuery()`'s methods will return itself as
+#' much as possible, unless the method is directly asking for information, e.g.
+#' `$selected()` or `$asTags()`.
+#'
+#' Internally, two important pieces of information are maintained: the root
+#' elements and the selected elements. The root tag environment will always
+#' point (after upgrading to a tag environment) to the original tag object
+#' provided to `tagQuery(tag=)`. However, the selected elements are a list of
+#' tag environments that update for every `$find(cssSelector)` call.  The
+#' selected elements are initialized to a list containing the `root` tag
+#' environment. All `tagQuery()` methods will act on the selected elements
+#' unless declared otherwise.
+#'
+#'
+#' @section Limitations:
+#'
+#'   `tagQuery()`s can **not** be used directly within typical `tag` locations.
+#'   An error should be thrown. Instead, please call `$asTags()` to retrieve the
+#'   tag structures of the selected tag elements or root element respectively.
+#'
+#' @param tags Any standard tag object or `tagList()`. If a `list()` or
+#'   `tagList()` is provided, a `tagList()` will be returned when calling
+#'   `$asTags()`.
+#' @return A `tagQuery()` object. The `tag` supplied will be considered the
+#'   `root` object. At the time of initialization, the `root` is also considered
+#'   the single selected item. If any selections are made, the selected elements
+#'   will be updated.
+#' @md
+#' @export
+tagQuery <- function(tags) {
+
+  tags <- flattenTagsRaw(tags %||% list())
+  tags <- asTagEnv(
+    wrapWithRootTag(tags)
+  )
```

Comment:
So suppose the following is run:

```R
y <- tagQuery(x).find("div")
z <- tagQuery(y$get(1))
```

This is a fairly common operation with jQuery. Does this mean that the whole tree will take a round trip to a lists and back to environments?

And then modifying `z` won't affect `y`?

I think it may make more sense to keep it as environments and allow modifications to `z` to affect `y`, but we need to think through it to make sure this doesn't have weird consequences.

Reply:
Another example:

```R
y <- tagQuery(x)
z <- tagQuery(y)
z$find("div")$addClass("foo")
```

I think it makes sense for this to modify `y` as well.

Reply:
Yes, we can have y & z share the same tag envs.

--------------------------

Let's say we do 

```r
y <- tagQuery(x)$find(".foo")
z <- tagQuery(y)
```

`y` now has selected elements within the tag environment of `x`. 

Question:
* Should `z` be the root of `y` with no selected elements? (Current implementation; Like passing DOM to another `tagQuery()`)
* Or should `z` be the root of `y` with the same selected elements of `y`? (Similar to jQuery;)

Reply:
If `tagQuery()` is called on a tag-env object, then traverse up all the parents until you hit the root, and build a new tagQuery object with that root, and the tag-env as the selection.

Reply:
Implemented in 28183fd